Mostly Autoconf cleanups

wferi opened the pull request:
A couple of small things I noticed during packaging the 1.0.1 release. They don't make a big difference for a small project like libqb, but I decided to offer them for the sake of consistency and future-proofing. 
Especially for pthreads, where I switched to the standard macro from the Autoconf Archive.
I couldn't fully test this on systems where `gethostbyname` or the socket functions aren't in libc, so some corrections might be needed if you support them for real.
Please consider incorporating these patches.

jnpkrn commented:
Minor nitpicks:
- could you please stick with common commit message prefixes
  and style that has been exercised lately
  (`build: [configure:]`, `Fix typos:`)?
- it's entirely fine to merge the single-change spelling fixes
- the first commit's message contains typo on its own (`defualt`)

jnpkrn commented on configure.ac:
```diff
@@ -466,10 +466,8 @@ AC_ARG_ENABLE([coverage],
 
-if test x"$with_check" = xyes; then
 AC_ARG_ENABLE([syslog-tests],
   [AS_HELP_STRING([--enable-syslog-tests],[build and run syslog tests])])
```
Can you advise as to whether this conditional is sensible, please?
My original intention was to lay an egg as carefully as possible because
libtool warns that `foo_LDFLAGS = -module` is not portable.

Any idea of having this portability tested in some minimalistic way
and turning "syslog-tests" on automatically if positive?

Or is it more a spurious warning of libtool?

jnpkrn replied:
Hm, cannot reproduce that easily, it might have been on FreeBSD 10.2 VM.

jnpkrn replied:
Reproduced on Linux box, but had to go back in history:

```
$ make distclean
$ git co -b syslog-test  642f74d
$ ./autogen.sh && ./configure --enable-syslog-tests && make check
...
*** Warning: Linking the executable log.test against the loadable module
*** _syslog_override.so is not portable!
...
```

Hard to say what resolved this.

wferi replied:
ebcff547bc47a6a3c195f6eb9d839ffed60a7381 did. Looks like portability is a concern for installed artifacts only. According to [SO](http://stackoverflow.com/questions/7619607/how-portable-is-linking-executables-against-loadable-modules) and the included [link](http://stackoverflow.com/questions/2339679/what-are-the-differences-between-so-and-dylib-on-osx), only Mac OS X is affected.

jnpkrn replied:
Yep, just have discovered that commit too; nice to have this positive
side effect :-)

In the new light, I would suggest dropping `--enable-syslog-tests`
unconditionally when the test suite is enabled as a whole.
I intend to do it subsequently on top of your PR, but feel free to drop
it right now without attempting to rectify anything related needlessly.

wferi commented:
I implemented your nits and reordered that patches. The last two (GitHub doesn't show them in ancestry order) may not be necessary if you intend to drop the `--enable-syslog-tests` option altogether. I can take them back if you prefer.

jnpkrn commented:
@wferi please have a look at #241 if it works for you

wferi commented:
#241 works for me perfectly, thanks.
Closing this.

jnpkrn commented:
On 08/12/16 13:30 -0800, wferi wrote:
> #241 works for me perfectly, thanks.
> Closing this.

I thank you for the exemplary cleanups, good work.

-- 
Jan (Poki)

jnpkrn commented on configure.ac:
```diff
@@ -108,7 +108,10 @@ dnl it will always be "none needed", but it is not true
+AX_SAVE_FLAGS
+AC_SEARCH_LIBS([dlopen],[dl],,[AC_MSG_ERROR([cannot find dlopen() function])])
+AC_SUBST([dlopen_LIBS],[$LIBS])
+AX_RESTORE_FLAGS
```
@wferi, discussed with @chrissie-c, she raised concerns about "keeping large(ish) chunks of other people's code in there just for simple things, it's just more stuff to keep up-to-date" and I must admit there's some grain of truth in it as all we really need here is just to back `LIBS`, nothing more, nothing less.  Would mind reopening this PR (if possible, I am not sure) and reflecting this comment, please?  Indeed, I can do it as a another follow up, but better to make history clean.

wferi replied:
Right there you don't even need to save `LIBS`, because it's empty. It's a little more interesting around the pthread function tests, where we also manipulate `CFLAGS` (which might be set by the user). Sure, all uses of `AX_SAVE/RESTORE_FLAGS` could easily be replaced by custom code. I decided to use them in this pull request, though, mostly for demo.
The main point is that they save you from thinking globally when doing local changes: you gain modularity and composability. You are free to add more involved checks (possibly requiring other flag changes) inside the save/restore pair or reorder them at will. They also reduce the room for error (I might be biased, but a couple of weeks ago I found the insidious `CFLAGS="-I/usr/include/security $CLFAGS"` and `CPPFLAGS="$ac_save_CFLAGS"` in the same `configure.ac` — after far too much head scratching).
Concerning the other point, all three macro files are straight from the Autoconf Archive. They seldom need updates, and when they do, it's a simple replace. Another option is requiring the archive to be installed for building libqb from Git.

jnpkrn commented:
On 09/12/16 13:12 -0800, wferi wrote:
> wferi commented on this pull request.
> 
>> @@ -108,7 +108,10 @@ dnl it will always be "none needed", but it is not true
>  dnl when linking libraries. Looks like a bug.
>  AC_SEARCH_LIBS([pthread_create], [pthread])
>  AC_SEARCH_LIBS([mq_open], [rt])
> -AC_SEARCH_LIBS([dlopen], [dl])
> +AX_SAVE_FLAGS
> +AC_SEARCH_LIBS([dlopen],[dl],,[AC_MSG_ERROR([cannot find dlopen() function])])
> +AC_SUBST([dlopen_LIBS],[$LIBS])
> +AX_RESTORE_FLAGS
> 
> Right there you don't even need to save `LIBS`, because it's empty.
> It's a little more interesting around the pthread function tests,
> where we also manipulate `CFLAGS` (which might be set by the user).

Ah, next time, I will not forward anything semi-unconsciously (for
some reason, I went by just the easy cases).

> Sure, all uses of `AX_SAVE/RESTORE_FLAGS` could easily be replaced
> by custom code. I decided to use them in this pull request, though,
> mostly for demo.
> The main point is that they save you from thinking globally when
> doing local changes: you gain modularity and composability. You are
> free to add more involved checks (possibly requiring other flag
> changes) inside the save/restore pair or reorder them at will. They
> also reduce the room for error (I might be biased, but a couple of
> weeks ago I found the insidious `CFLAGS="-I/usr/include/security
> $CLFAGS"` and `CPPFLAGS="$ac_save_CFLAGS"` in the same
> `configure.ac` — after far too much head scratching).

Yes, the ability to offload some tasks to well-tested, self-contained
solution, is appealing.  No doubts about that.

> Concerning the other point, all three macro files are straight from
> the Autoconf Archive. They seldom need updates, and when they do,
> it's a simple replace. Another option is requiring the archive to be
> installed for building libqb from Git.

That would just complicate consumption, in-tree is OK I believe.

I guess I let Chrissie speak for herself.


Btw. if there's anything to be done about the commits, I would have
a tiny wish (feel free to ignore if the PR would eventually be
accepted as is):
"configure: restrict pthreads to where it's actually needed" commit
could mention that mq_open is no longer relevant beyond commit
70a9623a48a7cc30ec079aa2762a35a4b7fc7add.

Thanks for bearing with me (us).

-- 
Jan (Poki)

wferi commented:
Jan Pokorný <notifications@github.com> writes:

> "configure: restrict pthreads to where it's actually needed" commit
> could mention that mq_open is no longer relevant beyond commit
> 70a9623a48a7cc30ec079aa2762a35a4b7fc7add.

True; I added this note and dropped the last two commits reorganizing
--enable-syslog-tests, since you'll remove that anyway.

> Thanks for bearing with me (us).

Come on, I actually count on you being picky and careful.  I've only got
a very narrow view of the project and zero experience with maintaining it.
-- 
Regards,
Feri

wferi commented:
... and the Travis checks failed. On code which is exactly the same as 2908242 in #241. Is this failure genuine? Or maybe the `ipc.test` has some race? The HPPA buildd failed [differently](https://buildd.debian.org/status/fetch.php?pkg=libqb&arch=hppa&ver=1.0.1-1&stamp=1481130367). (Hurd and kFreeBSD also fail some tests.)

jnpkrn commented:
On 12/12/16 03:14 -0800, wferi wrote:
> ... and the Travis checks failed. On code which is exactly the same
> as 2908242 in #241. Is this failure genuine? Or maybe the `ipc.test`
> has some race?

Yes, I suspect something like this:
https://github.com/ClusterLabs/libqb/issues/234

> The HPPA buildd failed
> [differently](https://buildd.debian.org/status/fetch.php?pkg=libqb&arch=hppa&ver=1.0.1-1&stamp=1481130367).
> (Hurd and kFreeBSD also fail some tests.)

check_ipc.c:1007:E:ipc_stress_connections_shm:test_ipc_stress_connections_shm:0:
(after this point) Test timeout expired

^ just adjusting the timeout for that test might help here.

Anyway, updated #241 and it passed:
https://travis-ci.org/ClusterLabs/libqb/builds/183218749

-- 
Jan (Poki)

chrissie-c commented:
The timeout for the stress test has to be HUGE on some platforms. I think it's set to an hour or so even from *BSD

jnpkrn commented:
It now causes a following problem with EL6:
https://copr-be.cloud.fedoraproject.org/results/%40ClusterLabs/devel/epel-6-x86_64/00490031-libqb/build.log.gz

```
autoreconf: running: /usr/bin/autoconf
configure:19163: error: possibly undefined macro: AS_VAR_COPY
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/bin/autoconf failed with exit status: 1
```

Note that I already encountered something similar with pacemaker,
where solution was to use `AS_VAR_GET` instead:

https://github.com/ClusterLabs/pacemaker/commit/36f7196f0bf656f95d24ce826ca13b2d1c697dbe#diff-a5522b90382504dc530177f63ced5209R46

@wferi, if you are OK with such a change, I'll go ahead along with
resetting serial to minimal value (currently unsure if 0 or 1).

wferi commented:
@jnpkrn, what do you think about introducing a "legacy" M4 file like for example https://github.com/varnishcache/varnish-cache/blob/master/varnish-legacy.m4? Definitions of missing macros could be put there as needed in the future. We could start with something like
~~~
# For compatibility with autoconf < 2.63b
m4_ifndef([AS_VAR_COPY],
  [m4_define([AS_VAR_COPY],
     [AS_LITERAL_IF([$1[]$2], [$1=$$2], [eval $1=\$$2])])])
~~~
which is basically a conditionalized copy of the `AS_VAR_COPY` definition in `/usr/share/autoconf/m4sugar/m4sh.m4`.

jnpkrn commented:
On 18/12/16 02:01 -0800, wferi wrote:
> @jnpkrn, what do you think about introducing a "legacy" M4 file like

Sounds reasonable, though it needs to be tested if it would be indeed
the "fixed point" for ~el6 already or more and more macros would have
to be introduced like this, which is something we would rather avoid.

Will have a look early in 2017, holiday season for me till then
(in case you have it similar, enjoy).

-- 
Jan (Poki)

wferi commented:
Sure we can return to this next year.
I'm already on holiday.  Merry Christmas!
-- 
Feri

wferi commented:
Thanks for the merge, @jnpkrn.
How did you solve the problem of `AS_VAR_COPY` being undefined under EL6 in the end?

jnpkrn commented:
On 31/05/17 07:53 -0700, wferi wrote:
> How did you solve the problem of `AS_VAR_COPY` being undefined under
> EL6 in the end?

It rests on the back burner so far, will hopefully get to is soon.

-- 
Jan (Poki)

inouekazu commented:
> wferi commented on 31 May
> How did you solve the problem of AS_VAR_COPY being undefined under EL6 in the end?

I made a patch https://github.com/inouekazu/libqb/commit/4b3c4c7d447a5773d286c7890f3d4136fae33653. how is it?

wferi commented:
> I made a patch inouekazu/libqb@4b3c4c7. how is it?

Looks fine, but is it enough to fix the EL6 build, as @jnpkrn queried?

inouekazu commented:
> Looks fine, but is it enough to fix the EL6 build, as @jnpkrn queried?

Yes. I checked with RHEL 6.9.
```
[root@rhel69-1 ~]# cat /etc/redhat-release
Red Hat Enterprise Linux Server release 6.9 (Santiago)

[root@rhel69-1 ~]# git clone https://github.com/ClusterLabs/libqb
[root@rhel69-1 ~]# cd libqb

[root@rhel69-1 libqb]# vi m4/ax_compat.m4
# For compatibility with autoconf < 2.63b
m4_ifndef([AS_VAR_COPY],
  [AC_DEFUN([AS_VAR_COPY],
     [AS_LITERAL_IF([$1[]$2], [$1=$$2], [eval $1=\$$2])])])

[root@rhel69-1 libqb]# ./autogen.sh
(snip)
examples/Makefile.am: installing `build-aux/depcomp'
autoreconf: Leaving directory `.'
Now run ./configure and make

[root@rhel69-1 libqb]# ./configure
(snip)

libqb configuration:
  Version                  = 1.0.2.18-333f
  Prefix                   = /usr
  Executables              = ${exec_prefix}/sbin
  Man pages                = ${datarootdir}/man
  Doc dir                  = ${datarootdir}/doc/${PACKAGE_TARNAME}
  Libraries                = /usr/lib64
  Header files             = ${prefix}/include
  Arch-independent files   = ${datarootdir}
  State information        = /var
  System configuration     = /etc
  SOCKETDIR                = /var/run
  Features                 =  epoll gcc__sync attribute-section

libqb build info:
  Optimization             =
  Debug options            =
  Extra compiler warnings  =
  Env. defined CFLAG       = -g -O2
  Env. defined CPPFLAGS    =
  Env. defined LDFLAGS     =
  ANSI defined CPPFLAGS    =
  Coverage     CFLAGS      =
  Coverage     LDFLAGS     =
  Fatal War.   CFLAGS      =
  Final        CFLAGS      = -g -O2       -Wall -Wextra -Wunused -Wshadow -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes -Wpointer-arith -Wwrite-strings -Wcast-align -Wbad-function-cast -Wmissing-format-attribute -Wfloat-equal -Wformat=2 -Woverlength-strings -Winit-self -Wuninitialized -Wunknown-pragmas -Wno-unused-parameter -Wno-format-nonliteral
  Final        CPPFLAGS    =
  Final        LDFLAGS     =
[root@rhel69-1 libqb]#
```
